src/OrchardCore.Modules/OrchardCore.Workflows/Controllers/WorkflowTypeController.cs

```diff
@@ -125,10 +125,19 @@ public async Task<IActionResult> Index(WorkflowTypeIndexOptions options, PagerPa
                 .ListAsync();
 
             var workflowTypeIds = workflowTypes.Select(x => x.WorkflowTypeId).ToList();
-            var workflowGroups = (await _session.QueryIndex<WorkflowIndex>(x => x.WorkflowTypeId.IsIn(workflowTypeIds))
-                .ListAsync())
-                .GroupBy(x => x.WorkflowTypeId)
-                .ToDictionary(x => x.Key);
+
+            // Changing the workflowTypeIds list to array is needed to be able to modify the original list while going
+            // through it.
+            foreach (var workflowTypeId in workflowTypeIds.ToArray())
```

Comment:
I presume if they are in the memory as a List it could be modified too

Reply:
Right, let us check if we can refactor this instead hit the database query many times

Reply:
What if we use

```csharp
var workflowInstanceToBeRemoved = await _session
.QueryIndex<WorkflowIndex>(index => !workflowTypeIds.Contains(index.WorkflowTypeId))
.ListAsync();
```

Then remove the instances from `workflowTypeIds`

Reply:
Contains won't work in this query. But we could use `.IsNotIn` which looks definitely better and it creates only one query to the database but it has to go through the whole table and return unnecessary data. And that brings back the original problem here.

The only information we need here is if there is a WorkflowIndex with the given workflowtypeid or not. So we are checking if it exists, separately for each workflow type id. If it is possible to check existence in a single query with YesSql, then that would be a better solution.


Something like 
```sql
SELECT
  SUM(CASE WHEN WorkflowTypeId = '4vn35yvjtmrft5ng3pedsexb5r' THEN 1 ELSE 0 END) AS Count_id1,
  SUM(CASE WHEN WorkflowTypeId = '40ka9pz8q3d0945vg5v4yd08nq' THEN 1 ELSE 0 END) AS Count_id2,
  SUM(CASE WHEN WorkflowTypeId = '4t07avan1z089vmk9eyd2mchx0' THEN 1 ELSE 0 END) AS Count_id3,
  SUM(CASE WHEN WorkflowTypeId = '4qyjq7q61fehhvy5frce3wrvmb' THEN 1 ELSE 0 END) AS Count_id4,
  SUM(CASE WHEN WorkflowTypeId = '4syp57rcmh4mp5shfr5s24zedw' THEN 1 ELSE 0 END) AS Count_id5
FROM WorkflowIndex;
```
would give back all the separate counts for each workflow type, but I'm not sure if we want this or it would make sense to make this possible with YesSql.

Reply:
> Contains won't work in this query

It's not support, right? I forgot I'm a LINQ guy but YesSql doesn't support all the operators :)

Reply:
Yeah it does not support it, so instead contains, we need to use `IsIn`.

Reply:
You could use an inverted from end to start for loop to avoid `ToArray()` call and you could also use `RemoveAt()` in this case making the code optimized.
This is just a suggestion

Reply:
I'm not sure if this is that big of a deal that it would require to do such loops.

Reply:
@hishamco let me know if it is okay like how it is currently

Reply:
I think Seb will disagree in terms of PERF :)

Reply:
Okay, if there is a better way to do it I'm glad to do it.

Note the current logic, what the original issue is about, is loading a workflow page with over half a million workflow instances within multiple seconds (6-30+) depending on if it is a fresh load in Azure environment. Locally it loads within 6-15 seconds. 
With this proposed fix it is around 20-30 **milliseconds** locally, so performance wise this is superior.

Reply:
I might try to replace the loop with n SQL queries into a single query if it's possible, we also might need to have a benchmark for this